src/transform/src/dataflow.rs

```diff
@@ -434,76 +448,382 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
-        indexes_by_view
-            .entry(sink_desc.1.from)
-            .or_insert_with(BTreeSet::new);
+    // This will be a mapping of
+    // (ids used by exports and objects to build) ->
+    // (arrangement keys and usage types on that id that have been explicitly requested by the MIR
+    // plans)
+    let mut index_reqs_by_id = BTreeMap::new();
+
+    // First, insert `sink_exports` and `index_exports` into `index_reqs_by_id` to account for the
+    // (currently only theoretical) possibility that we are using an index that we are also building
+    // ourselves.
+    for (_id, sink_desc) in dataflow.sink_exports.iter() {
+        index_reqs_by_id
+            .entry(sink_desc.from)
+            .or_insert_with(Vec::new);
     }
-    for (_, (index_desc, _)) in dataflow.index_exports.iter() {
-        indexes_by_view
+    for (_id, (index_desc, _)) in dataflow.index_exports.iter() {
+        index_reqs_by_id
             .entry(index_desc.on_id)
-            .or_insert_with(BTreeSet::new);
+            .or_insert_with(Vec::new);
     }
+
+    // Go through the MIR plans of `objects_to_build` and collect which arrangements are requested
+    // for which we also have an available index.
     for build_desc in dataflow.objects_to_build.iter_mut() {
-        build_desc
-            .plan
-            .as_inner_mut()
-            .visit_post(&mut |e| match e {
-                MirRelationExpr::Get {
-                    id: Id::Global(id), ..
-                } => {
-                    indexes_by_view.entry(*id).or_insert_with(BTreeSet::new);
-                }
-                MirRelationExpr::ArrangeBy { input, keys } => {
-                    if let MirRelationExpr::Get {
-                        id: Id::Global(id), ..
-                    } = &**input
-                    {
-                        for key in keys {
-                            if indexes.indexes_on(*id).find(|k| k == &key).is_some() {
-                                indexes_by_view.get_mut(id).unwrap().insert(key.clone());
-                            }
-                        }
-                    }
-                }
-                _ => {}
-            })?;
+        CollectIndexRequests::new(indexes, &mut index_reqs_by_id)
+            .collect_index_reqs(build_desc.plan.as_inner_mut())?;
     }
-    for (id, keys) in indexes_by_view.iter_mut() {
-        if keys.is_empty() {
-            // If a dataflow uses a view that has indexes, but it has not
-            // chosen to use any specific index of that view, pick an arbitrary
-            // index to read from.
+
+    // Add full index scans, i.e., where an index exists, but either no specific key was requested
+    // or the requested key didn't match an existing index in the above code.
+    for (id, requests_with_matching_indexes) in index_reqs_by_id.iter_mut() {
+        if requests_with_matching_indexes.is_empty() {
+            // Pick an arbitrary index to be fully scanned.
```

Comment:
How do we insure that the random pick here is consistent with the pick that we are going to do during MIR ⇒ LIR lowering or rendering?

Reply:
The MIR ⇒ LIR lowering or rendering has only the indexes to choose from that remain after `prune_and_annotate_dataflow_index_imports`. In the lowering, [this](https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/plan/mod.rs#L1793-L1822) is the code that goes through index imports and adds them to the `arrangements` map, and then it's used [here](https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/plan/mod.rs#L1091) when lowering a Get.

So, the burden for making good choices is fully on `prune_and_annotate_dataflow_index_imports`. For Joins, the MIR ArrangeBys are very clear about which keys will be needed. And for cases where no specific key was requested, the lowering is also just [picking](https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/plan/mod.rs#L1130) `arbitrary_arrangement`.